questions/249/explanation.md

```diff
@@ -1,11 +1,26 @@
 `'0'` is a character literal, with type `char`. (The value of `'0'` is actually implementation defined, but will typically be `48`.) This value is then promoted to an `int`, and stored in `a`.
 
-We then take a reference `b` to `a`. But `b` is a `char` reference, not an `int` reference, which means they are not *reference related*. §[dcl.init.ref]¶4:
-> Given types “cv1 `T1`” and “cv2 `T2`”, “cv1 `T1`” is reference-related to “cv2 `T2`” if `T1` is the same type as `T2`, or `T1` is a base class of `T2`.
+We then take a reference `b` to `a`. But `b` is a `char` reference, not an `int` reference, which means they are not *reference related*.
 
-"cv1 `T1`" here being `char const`, and "cv2 `T2`" being `int` , so they're not related. Since they're not,  §[dcl.init.ref]¶5.2.2.2 applies:
+§[dcl.init.ref]¶4:
 
-> Otherwise, the initializer expression is implicitly converted to a prvalue of type “cv1 `T1`”. The temporary materialization conversion is applied and the reference is bound to the result.
+> Given types “*cv1* `T1`” and “*cv2* `T2`”, “*cv1* `T1`” is reference-related to “*cv2* `T2`” if `T1` is similar (§[conv.qual]) to `T2`, or `T1` is a base class of `T2`.
+
+§[conv.qual]¶2:
+
+> Two types `T1` and `T2` are similar if they have qualification-decompositions with the same *n* such that corresponding *P<sub>i</sub>* components are either the same or one is “array of *N<sub>i</sub>*” and the other is “array of unknown bound of”, and the types denoted by `U` are the same.
```

Comment:
Maybe replace the stuff about arrays here and below with "(...)"?

Reply:
I shortened the §[conv.qual]¶2 clause (since in our case we already fail in "`U` are the same", no matter what P<sub>i</sub> components are), but I feel it's not OK to omit P<sub>i</sub> components in the §[conv.qual]¶1 clause, since it's a term definition, and omitting some parts of it makes that definition incomplete.

An alternative solution is to simply paraphrase both clauses in our own words without quoting them.

We should also think about similar fixes in №313.

Reply:
I noticed another issue too, the sentence that decides whether we can bind directly talks about "reference-compatible", not "reference-related". So I don't think we need to discuss reference-relatedness at all. I tried another approach:

---

`'0'` is a character literal, with type `char`. (The value of `'0'` is actually implementation defined, but will typically be `48`.) This value is then promoted to an `int`, and stored in `a`.

We then take a reference `b` to `a`. But `b` is a `char` reference, not an `int` reference, so it can not bind directly to the `int`.

§[dcl.init.ref]¶5 has a lot of rules, but since we're not dealing with rvalues or classes, the part that decides whether we can bind directly is:

>  If the reference is an lvalue reference and the initializer expression is an lvalue (but is not a bit-field), and “`cv1 T1`” is reference-compatible with “`cv2 T2`”

§[dcl.init.ref]¶4 describes reference-compatible as

> “`cv1 T1`” is reference-compatible with “`cv2 T2`” if a prvalue of type “pointer to `cv2 T2`” can be converted to the type “pointer to `cv1 T1`” via a standard conversion sequence §[conv]

A pointer to one type can not be converted to a pointer of another type (except for to a pointer to `void`), so these are not reference-compatible. Since they're not, §[dcl.init.ref]¶5.4.2 applies:

> — Otherwise, the initializer expression is implicitly converted to a prvalue of type “`T1`”. The temporary materialization conversion is applied, considering the type of the prvalue to be “*cv1* `T1`”, and the reference is bound to the result.

So the initializer expression `a` is converted to a temporary `char const`, which `b` refers to.

Reply:
Good catch! Sorry I didn't notice that myself. Fixed now.